datafusion/physical-expr/src/equivalence.rs

```
@@ -106,14 +110,89 @@ impl<T: Eq + Clone + Hash> EquivalenceProperties<T> {
             }
             (None, None) => {
                 // adding new pairs
-                self.classes.push(EquivalentClass::<T>::new(
+                self.classes.push(EquivalentClass::<Column>::new(
                     new_conditions.0.clone(),
                     vec![new_conditions.1.clone()],
                 ));
             }
             _ => {}
         }
     }
+
+    /// Normalizes physical expression according to `EquivalentClass`es inside `self.classes`.
+    /// expression is replaced with `EquivalentClass::head` expression if it is among `EquivalentClass::others`.
+    pub fn normalize_expr(&self, expr: Arc<dyn PhysicalExpr>) -> Arc<dyn PhysicalExpr> {
+        expr.clone()
+            .transform(&|expr| {
+                let normalized_form =
+                    expr.as_any().downcast_ref::<Column>().and_then(|column| {
+                        for class in &self.classes {
+                            if class.contains(column) {
+                                return Some(Arc::new(class.head().clone()) as _);
+                            }
+                        }
+                        None
+                    });
+                Ok(if let Some(normalized_form) = normalized_form {
+                    Transformed::Yes(normalized_form)
+                } else {
+                    Transformed::No(expr)
+                })
+            })
+            .unwrap_or(expr)
```

Why did we silence the error coming from here?

---

If we can't normalize, we return original expression.